BIP39 implementation

atalw opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

This PR is a custom BIP39 implementation which closes #561. 

There are a couple of features that have been implemented
- Parse mnemonic string to `Mnenomic` type
- Generate `Mnemonic` from entropy
- Derive seed from Mnemonic (with and without passphrase)
- All language wordlists (with verification test to ensure they were untampered)
- Mnemonic test vectors from [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#test-vectors)
- Error handling

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

### Notes to the reviewers

I've tried to keep the function names the same as they were to maintain backward compatibility. There is however a file structure change in the favour of clarity.

I was initially using tabs so earlier commits may fail CI. I'll amend if needed.

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature
* [ ] I've updated `CHANGELOG.md`

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

vladimirfomene commented:
Hello maintainers, before @atalw wrote on that he had started working on this issue, I had already asked to work on the issue and Daniela gave me the go ahead. He then asked whether we could collaborate and I agreed. I proposed we connect on discord and find a way to work together. On Discord, I told him, I already had an implementation of this BIP (https://github.com/vladimirfomene/bip39) in my repos. I asked if he minds me work ing on it alone saying that this was my first contribution and I wanted it to be sizeable. He said he had also started working on it (https://github.com/atalw/bdk/commit/3f56f33bb8c702be4c8e897dae8490a4879e6217) and thought it will be more prudent to collaborate on it. Seeing that he had already started working on it, I decide to collaborate with him and two hours after he sends me this PR (https://github.com/atalw/bdk/pull/1). In the PR, he has added a new commit which to_seed function which I think is inspired from my BIP39 implementation (check seed.rs in repo).  This was frustrating for me because he said let's work on it together but he had already implemented all the functionality. Even more frustrating is that I noticed he had imported the sha2, pbkdf2 and hmac crates and followed the exact algorithm I followed to implement to_seed.  He was using the rust-bitcoin hashes for sha256 and had suddenly switched to using sha2 for that just like I did in my implementation. After seeing this, I decided not to collaborate with him and reached out to @danielabrozzoni about the issue. She encouraged me to continue working on my patch and try to open a PR before @atalw . I continued working on my patch, and currently have an implementation that I'm testing (https://github.com/vladimirfomene/bdk/tree/implement-bip-39). If this PR gets merged, I will like to be added as a co-author. Also, why won't my PR be given preference given that I was the first to take up the issue? I understand how this is not easy for you as maintainers. Thank you for the work you do! Find attached my conversation with @atalw. 
![conversation1](https://user-images.githubusercontent.com/11140070/168558730-7e643c2a-df74-42c9-bc56-17feefd092cd.png)
![conversation2](https://user-images.githubusercontent.com/11140070/168558772-abb7e3ca-ca78-48d9-bf15-1ad8b9f83b09.png)

atalw commented:
Hey. Firstly, I'd like to say I'm really disappointed this is how our first contribution to BDK is going. Secondly, you mentioned you were willing to collaborate (after seeing the work I had already [done](https://github.com/atalw/bdk/commit/3f56f33bb8c702be4c8e897dae8490a4879e6217)) but in DM the first thing you asked is if you could work on it alone. The reason you're saying you didn't choose to collaborate was because I added a 20 line [function](https://github.com/atalw/bdk/pull/1/commits/cf28f48f48291150dfb9bb800c107c3ad75f4b26#diff-91ae8dd429c4d79c7f4a077d1b7e3cde95e7dcf611147e824dfba09f4fe3c3e2R124) which implemented all the functionality. I had opened a PR with a list of [next steps](https://github.com/atalw/bdk/pull/1#issuecomment-1125883233) and I tried to divide it amongst us but there was no response from you so I continued work.

I'm contributing to open-source to improve the projects and the ecosystem. I'm sorry that this situation was poorly handled. If there is no clear solution, to make things easier I could close this PR and find another issue to fix.

P.S the algorithm is similar because it is a BIP. I think there is essentially 1 way to do it with minor differences in coding style. Also I tried to use the rust-bitcoin hash functions (so that we wouldn't have to import a new crate) but `pbkdf2` was not accepting it.

vladimirfomene commented on Cargo.toml:
```diff
@@ -33,8 +33,11 @@ rocksdb = { version = "0.14", default-features = false, features = ["snappy"], o
 cc = { version = ">=1.0.64", optional = true }
 socks = { version = "0.3", optional = true }
 lazy_static = { version = "1.4", optional = true }
+pbkdf2 = { version = "0.10", optional = true }
+unicode-normalization = { version = "0.1.19", optional = true }
+sha2 = { version = "0.10.2", optional = true }
+hmac = { version = "0.12.1", optional = true }
```
You may want to use rust-bitcoin for sha256, sha512 and Hmac as they are already implemented in the project. As for pbkdf2, maybe you can consider porting the implementation in [rust-bip39](https://github.com/rust-bitcoin/rust-bip39/blob/master/src/pbkdf2.rs) while providing the necessary reference following the crate's library.

atalw replied:
I tried using them but `pbkfd2` wasn't accepting the rust-bitcoin `hmac` and `sha512` traits. Implementing `pbkdf2` should fix the issue.

As for porting `pbkdf2`, we could look into that. My only concern is that it may lead to bloat, especially if we simply copy it without understanding. If there are issues going forward, it will need to be maintained as well. So there's a tradeoff.

danielabrozzoni replied:
I spoke with @afilini and agreed that it's better if we port `pbkdf2` into BDK, as rust-bip39 is doing.

notmandatory commented:
My 2 cents on this is to remind everyone that reviewing someones PR and giving good feed back is as important, if not more important, than writing the actual code. Also that it's not the end of the world if two people write up code for the same issue, in fact it can be a good thing if they can discuss and compare their two solutions and find a way to collaborate on a single PR. Even if one of the two PRs isn't merged the knowledge gained by writing it can be a great way to learn the problem well enough to give good review feedback. Ideally you'll be able to work it out among your selves in a congenial way but worst case there are plenty of issues that need eyes on them and we can find  you different issues to work on.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+            };
+
+            indices.push(idx);
+        }
```
After calculating the indices for every word, you may want  to recover the entropy, then calculate the checksum to compare it against the binary representation of mnemonic. In case, the calculated checksum does not match the checksum in the binary representation of the mnemonic, you will want to throw an error.

atalw replied:
That's a good idea. Would you like to pick this up?

vladimirfomene replied:
Sorry for the late reply. I will work on it

danielabrozzoni commented:
> Firstly, I'd like to say I'm really disappointed this is how our first contribution to BDK is going.

Well, I'm kinda disappointed you couldn't sort this by yourselves, but ¯\_(ツ)_/¯

Since you've _kinda_ collaborated, I suggested the co-authoring solution. @atalw can you please squash your commits and put @vladimirfomene as a co-author? (See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors)

I'll review the PR soon, can you please fix the CI in the meantine? It's failing on `cargo clippy`:
```
error: manual `!RangeInclusive::contains` implementation
   --> src/keys/bip39/mod.rs:139:12
    |
139 |         if ent_bits < MIN_ENTROPY_BITS || ent_bits > MAX_ENTROPY_BITS {
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `!(MIN_ENTROPY_BITS..=MAX_ENTROPY_BITS).contains(&ent_bits)`
    |
    = note: `-D clippy::manual-range-contains` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_range_contains

error: use of `unwrap_or` followed by a function call
   --> src/keys/bip39/mod.rs:189:58
    |
189 |         let salt = ("mnemonic".to_string() + &passphrase.unwrap_or("".to_string()))
    |                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(|| "".to_string())`
    |
    = note: `-D clippy::or-fun-call` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

danielabrozzoni commented on Cargo.toml:
```diff
@@ -33,8 +33,11 @@ rocksdb = { version = "0.14", default-features = false, features = ["snappy"], o
 lazy_static = { version = "1.4", optional = true }
+pbkdf2 = { version = "0.10", optional = true }
+unicode-normalization = { version = "0.1.19", optional = true }
+sha2 = { version = "0.10.2", optional = true }
+hmac = { version = "0.12.1", optional = true }
```
Minimizing dependencies is important, you don't need those two: hmac and sha512 are already implemented in bitcoin_hashes, which is a dependency of BDK (re-exported by rust-bitcoin)

atalw replied:
I tried using them but `pbkfd2` wasn't accepting the rust-bitcoin `hmac` and `sha512` traits. Implementing `pbkdf2` should fix the issue.

danielabrozzoni commented on Cargo.toml:
```diff
@@ -33,8 +33,11 @@ rocksdb = { version = "0.14", default-features = false, features = ["snappy"], o
 cc = { version = ">=1.0.64", optional = true }
 socks = { version = "0.3", optional = true }
 lazy_static = { version = "1.4", optional = true }
+pbkdf2 = { version = "0.10", optional = true }
```
The MSRV of pbkdf2 is 1.57, ours is 1.56, maybe we should consider implementing our own pbkdf2 as rust-bip39 is doing? @afilini what do you think?

evanlinjin replied:
Are we still planning on having our own implementation of `pbkdf2`? If @atalw , does not mind, I can try pick this up?

atalw replied:
There is an agreement that `pbkdf2` should be implemented within BDK (refer [here](https://github.com/bitcoindevkit/bdk/pull/607#discussion_r897943179)). And yeah sure, feel free to implement it. I'm caught up with some other work and can only do it after 2 weeks.

vladimirfomene replied:
Go for it @evanlinjin !

evanlinjin replied:
Thank you @atalw and @vladimirfomene ! I've done the implementation here: #644 . Please let me know what I can improve on!

danielabrozzoni commented on Cargo.toml:
```diff
@@ -59,7 +62,7 @@ sqlite-bundled = ["sqlite", "rusqlite/bundled"]
 key-value-db = ["sled"]
 all-keys = ["keys-bip39"]
-keys-bip39 = ["bip39"]
+keys-bip39 = ["pbkdf2", "unicode-normalization", "sha2", "hmac"]
```
You should probably have different features for the different languages, as most of the people will use english only anyways

atalw replied:
Yeah, agreed. I'll add that.

atalw replied:
Added language feature flags. English is enabled by default.

danielabrozzoni reviewed (changes requested):
I have a failure locally:
```
test src/keys/mod.rs - keys::DerivableKey::into_extended_key (line 454) ... FAILED
---- src/keys/mod.rs - keys::DerivableKey::into_extended_key (line 454) stdout ----
error[E0277]: the trait bound `bdk::keys::bip39::Error: StdError` is not satisfied
  --> src/keys/mod.rs:464:6
   |
12 |     )?
   |      ^ the trait `StdError` is not implemented for `bdk::keys::bip39::Error`
   |
   = note: required because of the requirements on the impl of `From<bdk::keys::bip39::Error>` for `Box<dyn StdError>`
   = note: required because of the requirements on the impl of `FromResidual<Result<Infallible, bdk::keys::bip39::Error>>` for `Result<(), Box<dyn StdError>>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```
Can you fix?

Also, this is quite difficult to review as the commit history is not super clean... Can you please turn your 10 commits into 2:
- One MOVEONLY commit, that only moves all the old code from src/keys/bip39.rs to src/keys/bip39/mod.rs without changing the functionalities (it should look like this: https://github.com/bitcoin/bitcoin/commit/b36e738285b111a184c67fd845404850c34333a3)
- One commit where you actually add all the functionalities

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+            let mnemonic = vector[1];
+            let seed = Vec::from_hex(vector[2]).unwrap();
+            let xprivkey = vector[3];
+
```
You may want  to test the `parse_in` method here. You could do something like this:
```suggestion
assert_eq!(parse_in(Language::English, mnemonic).unwrap(), from_entropy_in(entropy).unwrap(), "failed mnemonic: {}", mnemonic);
```

atalw replied:
Added.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+    }
+
+    /// Convert a mnemonic to a seed with an optional passphrase
+    fn to_seed(&self, passphrase: Option<String>) -> [u8; 64] {
```
It might not be a good idea to change the type of `passphrase` as that has the potential of breaking code which consumes this method.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+    }
+
+    /// Convert a mnemonic to a seed with an optional passphrase
+    fn to_seed(&self, passphrase: Option<String>) -> [u8; 64] {
```
It might not be a good idea to change the type of `passphrase` from `&str` to `option<string>` as that has the potential of breaking code which consumes this method.

atalw replied:
It makes sense to have the passphrase as `Option` as it really is optional, so if a breaking change is okay we can go ahead with this.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+        assert_eq!(keys.len(), 1);
+        assert_eq!(networks.len(), 4);
+    }
+
```
I will add test to make sure that I'm catching failing condition for incorrect entropies.


```suggestion
//testing with entropy less than 128 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 15]).unwrap(), Err(Error::InvalidEntropyLength(120)));

//testing with entropy greater than 256 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 33]).unwrap(), Err(Error::InvalidEntropyLength(264)));

//testing entropy which is not a multiple of 32 bits
assert_eq!(from_entropy_in(Language::English, vec![b'0'; 31]).unwrap(), Err(Error::InvalidEntropyLength(248)));

```

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+        assert_eq!(keys.len(), 1);
+        assert_eq!(networks.len(), 4);
+    }
+
```
You might want to also add test for invalid mnemonic sentences.

```suggestion
//less than 12 words
assert_eq!(parse_in(Language::English, "advice cage absurd amount doctor acoustic avoid letter advice cage above").unwrap(), Err(Error::InvalidWordCount(11)));

//more than 24 words
assert_eq!(parse_in(Language::English, "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse achieve acid acoustic acquire across act action actor actress actual adapt add addict").unwrap(), Err(Error::InvalidWordCount(27)));

//not a multiple of six 
assert_eq!(parse_in(Language::English, "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse").unwrap(), Err(Error::InvalidWordCount(14)));

//invalid word
assert_eq!(parse_in(Language::English, "getter advice cage absurd amount doctor acoustic avoid letter advice cage above").unwrap(), Err(Error::InvalidWord(String::from("getter")));

//Once you implement the verification of the checksum in `parse_in` you might want to test that below here:
```

atalw replied:
While adding tests for incorrect entropies do you want to add these tests too?

vladimirfomene replied:
I'm not sure I understand you properly. Are you asking if it is necessary to test for these given that we are already testing incorrect entropies? If yes, then I think it is. I never saw a test for `parse_in` and I think it is necessary to have it.

atalw replied:
I think these tests should be there. I meant since you've written the code already, probably you should commit it as well.

vladimirfomene replied:
@atalw , I have added verification and also the tests I suggested above.

vladimirfomene commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,750 @@
+
+/// A BIP39 error
+#[derive(Debug)]
+pub enum Error {
```
While running the test on your patch, I had the following error:
![test-fail-bdk](https://user-images.githubusercontent.com/11140070/168699868-0608e31f-266a-4223-83e3-89dfb048b333.png)

This can be solved by implementing the `std::error::Error` trait on the error like so:
```suggestion
impl std::error::Error for Error{}
```

atalw replied:
Fixed.

vladimirfomene reviewed (changes requested):
This is a great start! You just need to add more test and implement a couple of verification in your code.

vladimirfomene reviewed (changes requested):
This is a great start! You just need to add more test and implement a couple of verification in your code.

atalw committed:
MOVEONLY: Move bip39.rs to bip39/mod.rs

atalw committed:
Custom BIP39 implementation

There are a couple of features that have been implemented in this
commit:
- Parse mnemonic string to Mnenomic type
- Generate Mnemonic from entropy
- Derive seed from Mnemonic (with and without passphrase)
- All language wordlists (with verification test to ensure they were
  untampered)
- Mnemonic test vectors from BIP39
- Error handling

Function names have mostly been kept the same to maintain backwards
compatability.

Co-authored-by: Vladimir Fomene <vladimirfomene@gmail.com>

atalw committed:
f - clippy fixes + rust fmt fix + failing test fix

atalw commented:
I've done a couple of things
1. Squashed all the commits into 2 commits – 1 MOVEONLY commit and 1 with all the functionality.
2. Added @vladimirfomene as a coauthor.
3. Fixed the clippy warnings and fixed the failing test.

I think it it's good for review now @danielabrozzoni. I'll address your comments and @vladimirfomene comments next.

atalw committed:
f - extended mnemonic tests to test `parse_in()`

atalw committed:
BIP39 - Add feature flags for mnemonic languages

atalw committed:
f - rust fmt

atalw committed:
f - rust fmt wordlists/mod.rs

vladimirfomene committed:
Verify checksum of mnemonic sentence

We are currently only checksum the
number of words in the mnemonic and their
validity. This implementation also checks the
checksum to make sure that it was generated
from the entropy of the mnemonic.

vladimirfomene committed:
Test with incorrect entropy size

Make sure `from_entropy_in` produces
error if length of entropy bits is
less than 128, greater than 256 and
not a multipe of 32.

vladimirfomene committed:
Test invalid mnemonic sentence

Throw error if mnemonic sentence is less
than 12 words or greater than 24 words or
number of words is not a multiple of six or
the contains a word not in wordlist or has
invalid checksum.

evanlinjin commented on src/keys/bip39/mod.rs:
```diff
@@ -0,0 +1,892 @@
+#[cfg_attr(docsrs, doc(cfg(feature = "keys-bip39")))]
+impl<Ctx: ScriptContext> DerivableKey<Ctx> for Seed {
+    fn into_extended_key(self) -> Result<ExtendedKey<Ctx>, KeyError> {
+        Ok(bip32::ExtendedPrivKey::new_master(Network::Bitcoin, &self[..])?.into())
```
Would it make sense to implement `DerivableKey` for `(bip39::Seed, Network)` instead of for `bip39::Seed` directly. This can apply for the other associated structures in this module as well.

I'm suggesting this because some people may want to derive testnet keys from the Mnemonic/Seed, and this hard-coding may lead to some confusion.


P.S. I bought this up in the [discord server](https://discord.com/channels/753336465005608961/978744259693916230/991207701692555324)

evanlinjin commented:
I propose that we close this PR and continue the conversation at #644 not sure if @atalw and @vladimirfomene will be okay with this?

danielabrozzoni commented:
@evanlinjin you might be able to push your code directly on @atalw's branch, so we don't lose all the comments when we change PR

evanlinjin commented:
> @evanlinjin you might be able to push your code directly on @atalw's branch, so we don't lose all the comments when we change PR

That's a good idea, I didn't know I had permission. I'll do some cherry-picking in a bit. 🍒 🌸

evanlinjin commented:
> @evanlinjin you might be able to push your code directly on @atalw's branch, so we don't lose all the comments when we change PR

It seems I do not have the right permissions to push code to his branch.
```
❯ git push atalw 
ERROR: Permission to atalw/bdk.git denied to evanlinjin.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

danielabrozzoni commented:
Then we can close this one and move to #644 :)
